File: backend/app/services/hrn.py
"""Motor de cálculo HRN — fonte da verdade.

HRN = LO × FE × DPH × NP, com fatores escolhidos das tabelas fixas da escala
ABNT NBR ISO/TR 14121-2. O cálculo e a classificação por faixa rodam aqui, no
backend; o frontend espelha as mesmas tabelas apenas para refletir em tempo real.

Princípio inegociável da metodologia: o HRN é a estimativa quantitativa do risco.
Ele nunca se mistura com a Categoria de segurança (NBR 14153 / PL ISO 13849-1),
que pertence a uma função de segurança — entidade separada de uma fase futura.
"""
from __future__ import annotations

from dataclasses import dataclass

# ---------------------------------------------------------------------------
# Tabelas de fatores: valor -> descrição (escala ABNT NBR ISO/TR 14121-2).
# A ordem é a da metodologia; os valores são a fonte da verdade.
# ---------------------------------------------------------------------------

LO: dict[float, str] = {
    0.033: "Quase impossível — só em circunstâncias extremas",
    0.5: "Altamente improvável",
    1: "Improvável, embora possível",
    1.5: "Possível",
    2: "Alguma chance",
    5: "Provável",
    8: "Provável a alta",
    10: "Muito provável",
    15: "Certeza — já documentado",
}

FE: dict[float, str] = {
    0.1: "Anualmente",
    0.2: "Mensalmente",
    1: "Semanalmente",
    1.5: "Diariamente",
    2.5: "Por hora",
    4: "Constantemente",
    5: "A cada minuto",
}

DPH: dict[float, str] = {
    0.1: "Arranhão / hematoma",
    0.5: "Corte / laceração leve",
    1: "Lesão leve, afastamento curto",
    2: "Afastamento — corte profundo, fratura simples",
    4: "Perda de 1 dedo / sequela menor",
    6: "Amputação parcial mão-pé · perda de visão (1 olho) · perda auditiva",
    8: "Amputação de membro · cegueira · fatalidade individual",
    10: "Múltiplas amputações · fatalidade",
    15: "Múltiplas fatalidades",
}

NP: dict[float, str] = {
    1: "1 pessoa",
    2: "2 pessoas",
    4: "3 a 7 pessoas",
    8: "8 a 15 pessoas",
    12: "Mais de 15 pessoas",
}

FACTOR_TABLES: dict[str, dict[float, str]] = {"lo": LO, "fe": FE, "dph": DPH, "np": NP}
FACTOR_NAMES: dict[str, str] = {
    "lo": "Probabilidade de ocorrência",
    "fe": "Frequência de exposição",
    "dph": "Grau do possível dano",
    "np": "Pessoas expostas",
}


# ---------------------------------------------------------------------------
# Faixas de classificação do HRN.
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Band:
    indice: int
    nome: str
    minimo: float
    maximo: float
    tratamento: str


# min inclusivo / max exclusivo, exceto a última (aberta). Um HRN igual ao limite
# sobe para a faixa seguinte — mesma convenção do protótipo (h >= min).
BANDS: list[Band] = [
    Band(0, "Aceitável", 0, 1, "Sem ação obrigatória"),
    Band(1, "Muito baixo", 1, 5, "Monitoramento, melhoria contínua"),
    Band(2, "Baixo", 5, 10, "Ação a planejar"),
    Band(3, "Significativo", 10, 50, "Ação necessária"),
    Band(4, "Alto", 50, 100, "Ação urgente"),
    Band(5, "Muito alto", 100, 500, "Ação imediata, condicionar operação"),
    Band(6, "Extremo", 500, 1000, "Parar a operação até mitigação"),
    Band(7, "Inaceitável", 1000, float("inf"), "Parar imediatamente"),
]


class HRNError(ValueError):
    """Fator fora da tabela fixa da escala."""


def _validate_factor(code: str, value: float) -> float:
    table = FACTOR_TABLES[code]
    # Comparação tolerante a float (0.033 etc.) contra as chaves da tabela.
    for key in table:
        if abs(float(key) - float(value)) < 1e-9:
            return float(key)
    raise HRNError(
        f"Valor {value!r} inválido para o fator {code.upper()}; "
        f"use um da escala: {sorted(table)}"
    )


def calcular_hrn(lo: float, fe: float, dph: float, np: float) -> float:
    """HRN = LO × FE × DPH × NP, validando cada fator contra a escala."""
    lo = _validate_factor("lo", lo)
    fe = _validate_factor("fe", fe)
    dph = _validate_factor("dph", dph)
    np = _validate_factor("np", np)
    return round(lo * fe * dph * np, 2)


def classificar(hrn: float) -> Band:
    """Retorna a faixa correta para um HRN (maior faixa cujo mínimo <= hrn)."""
    banda = BANDS[0]
    for b in BANDS:
        if hrn >= b.minimo:
            banda = b
    return banda


def classificar_nome(hrn: float) -> str:
    return classificar(hrn).nome


@dataclass
class AvaliacaoHRN:
    """Resultado completo da apreciação de um perigo (antes/depois + validação)."""

    hrn_atual: float
    classif_atual: str
    indice_atual: int
    hrn_pos: float | None
    classif_pos: str | None
    indice_pos: int | None
    queda_faixas: int | None
    nivel_validacao: str  # "ok" | "warn" | "bad" | "incompleto"
    mensagem_validacao: str
    fatores_pendentes: list[str]


def avaliar(
    *,
    lo: float,
    fe: float,
    dph: float,
    np: float,
    lo_pos: float | None = None,
    fe_pos: float | None = None,
    dph_pos: float | None = None,
    np_pos: float | None = None,
    justificativas: dict[str, str] | None = None,
) -> AvaliacaoHRN:
    """Avalia um perigo: HRN antes/depois, classificação e validações.

    As validações alertam, nunca bloqueiam (a decisão é do engenheiro):
    - o HRN depois deve cair pelo menos uma faixa;
    - o HRN depois não pode ser maior que o de antes (inconsistência);
    - cada fator deve ter justificativa — sem ela, o fator fica "pendente".
    """
    justificativas = justificativas or {}
    hrn_atual = calcular_hrn(lo, fe, dph, np)
    banda_atual = classificar(hrn_atual)

    pendentes = [
        c for c in ("lo", "fe", "dph", "np") if not (justificativas.get(c) or "").strip()
    ]

    tem_pos = None not in (lo_pos, fe_pos, dph_pos, np_pos)
    if not tem_pos:
        return AvaliacaoHRN(
            hrn_atual=hrn_atual,
            classif_atual=banda_atual.nome,
            indice_atual=banda_atual.indice,
            hrn_pos=None,
            classif_pos=None,
            indice_pos=None,
            queda_faixas=None,
            nivel_validacao="incompleto",
            mensagem_validacao="Apreciação incompleta — defina os fatores após as medidas.",
            fatores_pendentes=pendentes,
        )

    hrn_pos = calcular_hrn(lo_pos, fe_pos, dph_pos, np_pos)  # type: ignore[arg-type]
    banda_pos = classificar(hrn_pos)
    queda = banda_atual.indice - banda_pos.indice

    if banda_pos.indice > banda_atual.indice:
        nivel = "bad"
        msg = (
            "O risco aumentou após as medidas — o HRN de 'depois' não pode ser "
            "maior que o de 'antes'. Revise os fatores."
        )
    elif queda <= 0:
        nivel = "warn"
        msg = (
            "Sem redução de faixa. A apreciação só se completa quando o HRN cai "
            "pelo menos uma faixa; reforce as medidas de redução."
        )
    else:
        nivel = "ok"
        plural = "s" if queda > 1 else ""
        msg = (
            f"Risco reduzido em {queda} faixa{plural} — de '{banda_atual.nome}' "
            f"para '{banda_pos.nome}'."
        )

    return AvaliacaoHRN(
        hrn_atual=hrn_atual,
        classif_atual=banda_atual.nome,
        indice_atual=banda_atual.indice,
        hrn_pos=hrn_pos,
        classif_pos=banda_pos.nome,
        indice_pos=banda_pos.indice,
        queda_faixas=queda,
        nivel_validacao=nivel,
        mensagem_validacao=msg,
        fatores_pendentes=pendentes,
    )


def tabelas_para_frontend() -> dict:
    """Serializa as tabelas e faixas para o frontend espelhar a UI."""
    return {
        "fatores": {
            code: {
                "nome": FACTOR_NAMES[code],
                "opcoes": [{"valor": v, "descricao": d} for v, d in table.items()],
            }
            for code, table in FACTOR_TABLES.items()
        },
        "faixas": [
            {
                "indice": b.indice,
                "nome": b.nome,
                "min": b.minimo,
                "max": None if b.maximo == float("inf") else b.maximo,
                "tratamento": b.tratamento,
            }
            for b in BANDS
        ],
        "niveis_iso12100": [
            {"codigo": "M1", "descricao": "Projeto intrinsecamente seguro"},
            {"codigo": "M2", "descricao": "Proteção / dispositivo de segurança"},
            {"codigo": "M3", "descricao": "Informação para uso"},
            {"codigo": "M4", "descricao": "EPI"},
        ],
    }

File: frontend/src/lib/hrn.ts
// Espelho das tabelas e faixas do backend (app/services/hrn.py).
// O backend é a fonte da verdade; aqui replicamos só para o recálculo em tempo
// real na UI. Mantenha os valores idênticos aos do backend.

export type FactorCode = "lo" | "fe" | "dph" | "np";

export const FACTOR_NAMES: Record<FactorCode, string> = {
  lo: "Probabilidade de ocorrência",
  fe: "Frequência de exposição",
  dph: "Grau do possível dano",
  np: "Pessoas expostas",
};

export interface Opcao {
  valor: number;
  descricao: string;
}

export const TABELAS: Record<FactorCode, Opcao[]> = {
  lo: [
    { valor: 0.033, descricao: "Quase impossível — só em circunstâncias extremas" },
    { valor: 0.5, descricao: "Altamente improvável" },
    { valor: 1, descricao: "Improvável, embora possível" },
    { valor: 1.5, descricao: "Possível" },
    { valor: 2, descricao: "Alguma chance" },
    { valor: 5, descricao: "Provável" },
    { valor: 8, descricao: "Provável a alta" },
    { valor: 10, descricao: "Muito provável" },
    { valor: 15, descricao: "Certeza — já documentado" },
  ],
  fe: [
    { valor: 0.1, descricao: "Anualmente" },
    { valor: 0.2, descricao: "Mensalmente" },
    { valor: 1, descricao: "Semanalmente" },
    { valor: 1.5, descricao: "Diariamente" },
    { valor: 2.5, descricao: "Por hora" },
    { valor: 4, descricao: "Constantemente" },
    { valor: 5, descricao: "A cada minuto" },
  ],
  dph: [
    { valor: 0.1, descricao: "Arranhão / hematoma" },
    { valor: 0.5, descricao: "Corte / laceração leve" },
    { valor: 1, descricao: "Lesão leve, afastamento curto" },
    { valor: 2, descricao: "Afastamento — corte profundo, fratura simples" },
    { valor: 4, descricao: "Perda de 1 dedo / sequela menor" },
    { valor: 6, descricao: "Amputação parcial mão-pé · perda de visão (1 olho) · perda auditiva" },
    { valor: 8, descricao: "Amputação de membro · cegueira · fatalidade individual" },
    { valor: 10, descricao: "Múltiplas amputações · fatalidade" },
    { valor: 15, descricao: "Múltiplas fatalidades" },
  ],
  np: [
    { valor: 1, descricao: "1 pessoa" },
    { valor: 2, descricao: "2 pessoas" },
    { valor: 4, descricao: "3 a 7 pessoas" },
    { valor: 8, descricao: "8 a 15 pessoas" },
    { valor: 12, descricao: "Mais de 15 pessoas" },
  ],
};

export interface Band {
  indice: number;
  nome: string;
  min: number;
  max: number; // Infinity na última
  cor: string;
  txt: string;
  tratamento: string;
}

export const BANDS: Band[] = [
  { indice: 0, nome: "Aceitável", min: 0, max: 1, cor: "--b0", txt: "#fff", tratamento: "Sem ação obrigatória" },
  { indice: 1, nome: "Muito baixo", min: 1, max: 5, cor: "--b1", txt: "#fff", tratamento: "Monitoramento, melhoria contínua" },
  { indice: 2, nome: "Baixo", min: 5, max: 10, cor: "--b2", txt: "#16181D", tratamento: "Ação a planejar" },
  { indice: 3, nome: "Significativo", min: 10, max: 50, cor: "--b3", txt: "#16181D", tratamento: "Ação necessária" },
  { indice: 4, nome: "Alto", min: 50, max: 100, cor: "--b4", txt: "#16181D", tratamento: "Ação urgente" },
  { indice: 5, nome: "Muito alto", min: 100, max: 500, cor: "--b5", txt: "#fff", tratamento: "Ação imediata, condicionar operação" },
  { indice: 6, nome: "Extremo", min: 500, max: 1000, cor: "--b6", txt: "#fff", tratamento: "Parar a operação até mitigação" },
  { indice: 7, nome: "Inaceitável", min: 1000, max: Infinity, cor: "--b7", txt: "#fff", tratamento: "Parar imediatamente" },
];

export function bandIndex(h: number): number {
  let i = 0;
  for (let k = 0; k < BANDS.length; k++) if (h >= BANDS[k].min) i = k;
  return i;
}

export function classificar(h: number): Band {
  return BANDS[bandIndex(h)];
}

export function calcularHRN(lo: number, fe: number, dph: number, np: number): number {
  return Math.round(lo * fe * dph * np * 100) / 100;
}

export const fmtN = (n: number) => String(n).replace(".", ",");

export function fmtHRN(h: number): string {
  if (!isFinite(h)) return "—";
  const r = Math.round(h * 100) / 100;
  return fmtN(Number.isInteger(r) ? r : parseFloat(r.toFixed(2)));
}

// Posição (%) do marcador na barra de faixas — replica o protótipo (escala log
// dentro de cada faixa, 12,5% por faixa).
export function scalePos(h: number): number {
  const i = bandIndex(h);
  const w = 12.5;
  const base = i * w;
  const clamp = (v: number, a: number, b: number) => Math.max(a, Math.min(b, v));
  let f: number;
  if (i === 0) f = clamp(h / 1, 0, 1);
  else if (i === 7) f = clamp((Math.log10(h) - 3) / (Math.log10(5000) - 3), 0, 1);
  else {
    const lo = BANDS[i].min;
    const hi = BANDS[i].max;
    f = clamp((Math.log10(h) - Math.log10(lo)) / (Math.log10(hi) - Math.log10(lo)), 0, 1);
  }
  return base + f * w;
}

export const NIVEIS_ISO: [string, string][] = [
  ["M1", "Projeto intrinsecamente seguro"],
  ["M2", "Proteção / dispositivo de segurança"],
  ["M3", "Informação para uso"],
  ["M4", "EPI"],
];

export interface AvaliacaoLocal {
  hrnAntes: number;
  hrnDepois: number | null;
  bandAntes: Band;
  bandDepois: Band | null;
  queda: number | null;
  nivel: "ok" | "warn" | "bad" | "incompleto";
  mensagem: string;
}

export function avaliarLocal(
  antes: Record<FactorCode, number>,
  depois: Record<FactorCode, number> | null,
): AvaliacaoLocal {
  const hrnAntes = calcularHRN(antes.lo, antes.fe, antes.dph, antes.np);
  const bandAntes = classificar(hrnAntes);
  if (!depois) {
    return {
      hrnAntes, hrnDepois: null, bandAntes, bandDepois: null, queda: null,
      nivel: "incompleto",
      mensagem: "Apreciação incompleta — defina os fatores após as medidas.",
    };
  }
  const hrnDepois = calcularHRN(depois.lo, depois.fe, depois.dph, depois.np);
  const bandDepois = classificar(hrnDepois);
  const queda = bandAntes.indice - bandDepois.indice;
  let nivel: AvaliacaoLocal["nivel"];
  let mensagem: string;
  if (bandDepois.indice > bandAntes.indice) {
    nivel = "bad";
    mensagem = "O risco aumentou após as medidas — o HRN de 'depois' não pode ser maior que o de 'antes'. Revise os fatores.";
  } else if (queda <= 0) {
    nivel = "warn";
    mensagem = "Sem redução de faixa. A apreciação só se completa quando o HRN cai pelo menos uma faixa; reforce as medidas.";
  } else {
    nivel = "ok";
    mensagem = `Risco reduzido em ${queda} faixa${queda > 1 ? "s" : ""} — de "${bandAntes.nome}" para "${bandDepois.nome}".`;
  }
  return { hrnAntes, hrnDepois, bandAntes, bandDepois, queda, nivel, mensagem };
}
